pprof provider

bevzzz opened the pull request:
## Summary
`pprof` provider annotates functions/methods with CPU time / memory allocations associated with them.

## Motivation
The idea came to me as I was reading about profiling Go programs with [`pprof`](https://github.com/google/pprof). The provider consists of a simple Go parser and a minimal wrapper around `go tool pprof` / `pprof`. When called, it will search through the user's workspace to find a file matching the "pprof report" glob and collect statics for each function _declared_ in this file.

Below is the example of the annotations created for the loop finding algorithm from the [Profiling Go Programs](https://go.dev/blog/pprof) blog post.

| Annotations | Function |
|--------|--------|
| ![image](https://github.com/sourcegraph/openctx/assets/53943884/8bedf69f-6cb1-47c0-a9b3-9825d323e467) | ![Screenshot 2024-06-06 at 22 58 29](https://github.com/sourcegraph/openctx/assets/53943884/e9b7a399-90d3-444d-9c2f-793088ed2fa3) | 

The way I imagine it, such annotations would allow Cody to assist developers at tackling performance bottlenecks by making suggestions based on factual data (metrics). I still haven't figured out a way to make this context available to Cody -- that's something for tomorrow's pairing session.

## TODOs:

- [x] Include data from memory profile reports
- [x] Support standalone `pprof` (not as part of `go` binary)
- [x] Find binary from which the report was generated. Passing it as an argument to `pprof` unlocks some additional functionality and makes the output more precise.
-  ~~Make this info available to Cody~~ 
Upd: Not something we can do just now, needs a separate implementation in Cody.
- [x] Docs

Some additional features that I have in mind for it:
- Annotating "resource hungry"  lines using the detailed output of `pprof -list`
Upd: for now just passing the output of this command to `annotations.items.ai`
- Adding a preview of the function's call graph with `pprof svg`
- Support other languages: there are profiling tools which generate `pprof`-compatible reports for other languages (e.g. [`pprof-nodejs`](https://github.com/google/pprof-nodejs))

## Try it out

To use this provider you'll need to:
- Check out the branch
- Find a Go project and run a profiler on it
- Enable the `openctx/provider/pprof` provider and give it the report glob to look for
- Open a Go file (mind that not every file will have functions with CPU times significant enough to be included in `pprof`'s output)

bevzzz committed:
feat(pprof): initial commit

bevzzz committed:
feat(pprof): calculate Range for each func/method

bevzzz committed:
fix(pprof): escape special characters for pprof search

bevzzz committed:
wip(pprof): search pprof report and go binary

Before generating annotations for a file, we want to check if
1) go is installed (to use `go tool pprof`)
2) a pprof report that matches the configured glob exists in
one of the parent directories

bevzzz committed:
feat(pprof): get `pprof -top` for current package

This makes sure the annotations are only set for the most "resource
hungry" functions and not for each function in the package.

bevzzz committed:
fix(pprof/parser): espace all special characters in the string

bevzzz committed:
feat(pprof): annotate Go files

Putting it all together!

`profider-pprof` can now annotate Go source files with information about
the CPU time the functions/methods defined in it take.
This requires `go` to be installed (standalone `pprof` will be supported
too) and a report generated.

Some remarks:
- Bungling to --format=cjs until this fix
https://github.com/sourcegraph/openctx/pull/131 is merged
- _log.ts file is a temporary fixture for local development
- Lots of TODOs to handle

keegancsmith reviewed (commented):
nice. we have discussed a way to send over in mentions/items some editor context. But thinking a bit more about it, maybe we should add a way to mention the current files annotations (or selected lines annotations)? The annotation response has an item field which can then be used for context. cc @sqs @thenamankumar

keegancsmith commented on provider/pprof/index.ts:
```diff
@@ -0,0 +1,89 @@
+            return []
+        }
+
+        // let workspaceRoot: string | undefined = undefined
+        // const workspaceFolder = vscode.workspace.getWorkspaceFolder(vscode.Uri.parse(params.uri))
+        // if (workspaceFolder && workspaceFolder !== null) {
+        //     workspaceRoot = workspaceFolder.uri.path
+        // }
```
I see you have this commented out, but you still have an incantation in your esbuild which shows you want to do this. Did you have trouble getting this to work?

bevzzz commented:
> But thinking a bit more about it, maybe we should add a way to mention the current files annotations (or selected lines annotations)?

I'm thinking about looking at the Cody repo after I'm done with this PR, maybe that's something I could implement?

bevzzz committed:
test(pprof): enable all tests and simplify mocking

execSync does not return an error if the command
fails, so we check that its output is not "not found".

bevzzz committed:
fix(pprof): missing dot in default reportGlob

bevzzz committed:
chore(pprof): return no annotations for a test file

bevzzz committed:
chore(pprof): drop 'vscode' dependency, remove unused code

bevzzz committed:
ci: appease biome

RXminuS commented on provider/pprof/package.json:
```diff
@@ -0,0 +1,29 @@
+  "sideEffects": false,
+  "scripts": {
+    "build": "tsc --build",
+    "bundle": "tsc --build && esbuild --log-level=error --bundle --format=cjs --platform=node --external:vscode --outfile=dist/bundle.js index.ts",
```
Interesting, does this work for you with `--format=cjs` ? (I'm trying to do something similar but from what I can tell the `importHelper.ts` always assumes it's esm.

bevzzz replied:
It does, although it's a temporary patch until [this fix](https://github.com/sourcegraph/openctx/pull/131) is released.

You're right, `importHelper.ts` assumes the bundles are `esm`, but will still try to convert them to `cjs` if it's run on desktop. Since this bundle is already in `cjs` format it doesn't break things.

bevzzz replied:
Updated to `--format=esm`, the latest OpenCtx version can now compile it correctly.

bevzzz committed:
feat(pprof): support standalone pprof

Usually pprof is included in go distribution, but the extention should
be able to use either of the tools, whichever one's installed.

bevzzz committed:
feat(pprof): pass binary to pprof if found

bevzzz committed:
feat(pprof): match reportGlob to a full file path

This should allow users to specify directories where the pprof reports
should and shouldn't be taken from.

bevzzz committed:
fix(pprof): get units from `pprof -top` with >1 character

bevzzz committed:
feat(pprof): pass detailed breakdown (pprof list) to "ai" field in annotation's item

bevzzz committed:
refactor(pprof): store file's functions as a Record to minimize looping

In the `annotations()` we would previously loop over every function in a file, then loop over the function from `pprof top` to find information for the given function.
Since "top" results should remain ordered, it seems easier to return funcs as a Record<string, Func>

bevzzz committed:
chore(pprof): change annotation title

bevzzz committed:
feat(pprof): apply all provider settings

Previously reportGlob, binaryGlob, and all TopOptions
were ignored even if set in settings.json.

bevzzz committed:
chore: appease biome

tinvaan commented on provider/pprof/README.md:
```diff
@@ -0,0 +1,24 @@
+
+## Usage
+
+Add the following to your settings in any OpenCtx client:
```
Would this provider work in browser environments?

bevzzz replied:
No, I don't think so. The provider expects to find the pprof report somewhere on your local machine, which the others using the extension would probably not have on their machines.

I'm going to update the docs today 👌

bevzzz committed:
build(pprof): bundle to esm format

bevzzz committed:
fix(pprof): avoid breaking search loop prematurely

When searching for the Go binary, we assume that it will be in the same directory as the report, or in one of its child directories.
If we break the loop the moment we find the report, then we might not find the binary even if it's in the same directory but is alphabetically sorted after the report file.

bevzzz committed:
ci: fix formatting

bevzzz committed:
chore(pprof): write docs

bevzzz committed:
ci: appease biomi

bevzzz committed:
ci: run biome format --fix

bevzzz committed:
ci: fix more biome errors

bevzzz committed:
ci: appease biome
